lib/Dialect/FIRRTL/Transforms/MaterializeDebugInfo.cpp

```diff
@@ -0,0 +1,111 @@
+};
+} // namespace
+
+void MaterializeDebugInfoPass::runOnOperation() {
+  auto module = getOperation();
+  auto builder = OpBuilder::atBlockBegin(module.getBodyBlock());
+
+  // Create DI variables for each port.
+  for (const auto &[port, value] :
+       llvm::zip(module.getPorts(), module.getArguments())) {
+    materializeVariable(builder, port.name, value);
+  }
+
+  // Create DI variables for each declaration in the module body.
+  module.walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op).Case<WireOp, NodeOp, RegOp, RegResetOp>(
+        [&](auto op) {
+          builder.setInsertionPointAfter(op);
+          materializeVariable(builder, op.getNameAttr(), op.getResult());
+        });
+  });
+}
+
+/// Materialize debug variable ops for a value.
+void MaterializeDebugInfoPass::materializeVariable(OpBuilder &builder,
+                                                   StringAttr name,
+                                                   Value value) {
+  if (!name || isUselessName(name.getValue()))
+    return;
+  if (name.getValue().starts_with("_"))
+    return;
+  if (auto dbgValue = convertToDebugAggregates(builder, value))
+    builder.create<debug::VariableOp>(value.getLoc(), name, dbgValue);
+}
+
+/// Unpack all aggregates in a FIRRTL value and repack them as debug aggregates.
+/// For example, converts a FIRRTL vector `v` into `dbg.array [v[0],v[1],...]`.
+Value MaterializeDebugInfoPass::convertToDebugAggregates(OpBuilder &builder,
+                                                         Value value) {
+  return TypeSwitch<Type, Value>(value.getType())
+      .Case<BundleType>([&](auto type) {
+        SmallVector<Value> fields;
+        SmallVector<Attribute> names;
+        SmallVector<Operation *> subOps;
+        for (auto [index, element] : llvm::enumerate(type.getElements())) {
+          auto subOp = builder.create<SubfieldOp>(value.getLoc(), value, index);
+          subOps.push_back(subOp);
+          if (auto dbgValue = convertToDebugAggregates(builder, subOp)) {
+            fields.push_back(dbgValue);
+            names.push_back(element.name);
+          }
+        }
+        auto result = builder.create<debug::StructOp>(
+            value.getLoc(), fields, builder.getArrayAttr(names));
+        for (auto *subOp : subOps)
+          if (subOp->use_empty())
+            subOp->erase();
+        return result;
+      })
+      .Case<FVectorType>([&](auto type) -> Value {
+        SmallVector<Value> elements;
+        SmallVector<Operation *> subOps;
+        for (unsigned index = 0; index < type.getNumElements(); ++index) {
+          auto subOp = builder.create<SubindexOp>(value.getLoc(), value, index);
+          subOps.push_back(subOp);
+          if (auto dbgValue = convertToDebugAggregates(builder, subOp))
+            elements.push_back(dbgValue);
+        }
+        Value result;
+        if (!elements.empty())
+          result = builder.create<debug::ArrayOp>(value.getLoc(), elements);
```

Comment:
Is it possible that `elements` has less elements than `getNumElements()`? If so I think it's might be confusing to get values with incorrect indexes. If it's impossible to happen then can we make it assertion?

Reply:
Oh yeah, great catch. Just dropping elements will lead to incorrect indices. I don't think there will be cases in practice where `convertToDebugAggregates` works for some fields in the array and not for others. But it would be good to just not create the array in that case. Better not show any information than messed up indices. An alternative would be to have a `dbg.unknown/optimized/removed/lost` value to indicate values that can't reliably be reproduced.

To get started, let me just skip the entire array if debug aggregate conversion failed for any of its fields.

Reply:
Done!